aws/middleware/endpoints.go

```diff
@@ -0,0 +1,69 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+)
+
+// ResolveServiceEndpoint is a middleware that will resolve the service endpoint.
+type ResolveServiceEndpoint struct {
+	Resolver aws.EndpointResolver
+}
+
+// ID is a middleware identifier
+func (r ResolveServiceEndpoint) ID() string {
+	return "ResolveServiceEndpoint"
+}
+
+// HandleSerialize resolves the service endpoint and returns
+func (r ResolveServiceEndpoint) HandleSerialize(
+	ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler,
+) (out middleware.SerializeOutput, metadata middleware.Metadata, err error) {
+	req, ok := in.Request.(*smithyhttp.Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown transport type %T", in.Request)
+	}
+
+	var endpoint aws.Endpoint
+	endpoint, err = r.Resolver.ResolveEndpoint(GetEndpointID(ctx), GetRegion(ctx))
+	if err != nil {
+		return out, metadata, fmt.Errorf("failed to resolve service endpoint ")
+	}
+
+	req.URL, err = url.Parse(endpoint.URL)
+	if err != nil {
+		return out, metadata, fmt.Errorf("failed to parse endpoint URL: %w", err)
+	}
+
+	if len(endpoint.SigningName) > 0 && (len(GetSigningName(ctx)) == 0 || !endpoint.SigningNameDerived) {
+		ctx = SetSigningName(ctx, endpoint.SigningName)
+	}
+	ctx = SetSigningRegion(ctx, endpoint.SigningRegion)
+
+	return next.HandleSerialize(ctx, in)
```

Comment:
We'll probably want special behavior that handles the case where a endpoint contains a standard port, e.g. 80/443. These ports need to be stripped off the "host" header before signing, or ths signature might become invalid.

Reply:
https://github.com/aws/aws-sdk-go-v2/issues/606 to track this work.